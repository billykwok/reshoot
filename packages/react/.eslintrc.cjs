module.exports = {
  root: false,
  env: {
    browser: true,
    node: true,
    commonjs: true,
    es6: true,
    'jest/globals': true,
  },
  parser: '@babel/eslint-parser',
  parserOptions: {
    ecmaVersion: 'latest',
    sourceType: 'module',
    ecmaFeatures: { impliedStrict: true, jsx: true },
    babelOptions: { configFile: './babel.config.cjs' },
  },
  settings: {
    react: { version: 'detect' },
    'import/extensions': ['.js', '.jsx', '.json'],
    'import/resolver': { node: { moduleDirectory: ['node_modules'] } },
  },
  plugins: [
    'import',
    'prettier',
    'react',
    'react-hooks',
    'jsx-a11y',
    'testing-library',
    'jest',
  ],
  extends: [
    'eslint:recommended',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended',
    'prettier',
    'plugin:testing-library/react',
    'plugin:jest/recommended',
  ],
  rules: {
    'react/prop-types': 'off',
    'react/jsx-uses-react': 'off',
    'react/react-in-jsx-scope': 'off',
    'jsx-a11y/image-has-content': 'off',
    'jsx-a11y/heading-has-content': 'off',
    'jsx-a11y/anchor-has-content': 'off',
    'react-hooks/rules-of-hooks': 'error',
    'prefer-spread': 'off',
    'consistent-return': 'off',
    'max-len': 'off',
    'no-console': 'off',
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        ecmaVersion: 'latest',
        sourceType: 'module',
        ecmaFeatures: { impliedStrict: true, jsx: true },
        project: ['./tsconfig.json'],
      },
      settings: {
        react: { version: 'detect' },
        'import/extensions': ['.ts', '.tsx'],
        'import/resolver': { node: { moduleDirectory: ['node_modules'] } },
        'import/parsers': { '@typescript-eslint/parser': ['.ts', '.tsx'] },
      },
      plugins: [
        '@typescript-eslint',
        'import',
        'prettier',
        'react',
        'react-hooks',
        'jsx-a11y',
        'testing-library',
        'jest',
      ],
      extends: [
        'eslint:recommended',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'plugin:react/recommended',
        'plugin:jsx-a11y/recommended',
        'plugin:prettier/recommended',
        'prettier',
        'plugin:testing-library/react',
        'plugin:jest/recommended',
      ],
      rules: {
        '@typescript-eslint/no-var-requires': 'off',
        '@typescript-eslint/no-explicit-any': 'off',
        '@typescript-eslint/no-floating-promises': 'off',
        '@typescript-eslint/explicit-function-return-type': 'off',
        '@typescript-eslint/explicit-module-boundary-types': [
          'error',
          { allowArgumentsExplicitlyTypedAsAny: true },
        ],
        '@typescript-eslint/prefer-includes': 'off',
        'react/prop-types': 'off',
        'react/jsx-uses-react': 'off',
        'react/react-in-jsx-scope': 'off',
        'jsx-a11y/image-has-content': 'off',
        'jsx-a11y/heading-has-content': 'off',
        'jsx-a11y/anchor-has-content': 'off',
        'react-hooks/rules-of-hooks': 'error',
      },
    },
  ],
};
