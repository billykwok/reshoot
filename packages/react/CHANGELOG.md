# @reshoot/react

## 2.0.5

### Patch Changes

- a0288f4: fix: support hash character in inline options

## 2.0.4

### Patch Changes

- ec76336: fix: improve rollup setup and update dependencies
- Updated dependencies [ec76336]
  - @reshoot/types@2.0.3

## 2.0.3

### Patch Changes

- 0e54ec5: fix: fix style exports

## 2.0.2

### Patch Changes

- 510a7da: fix: fix incorrect extensions in package.json

## 2.0.1

### Patch Changes

- 226a185: fix: add node version restriction and fix rollup config
- Updated dependencies [226a185]
  - @reshoot/types@2.0.1

## 2.0.0

### Major Changes

- 8d162cb: Introducing Reshoot 2

  Reshoot has been rewritten from ground up to support more use cases. Please refer to README.md for details.

### Patch Changes

- Updated dependencies [8d162cb]
  - @reshoot/types@2.0.0
