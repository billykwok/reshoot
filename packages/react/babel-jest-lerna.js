module.exports = require('babel-jest').createTransformer({
  rootMode: 'upward',
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    [
      '@babel/preset-typescript',
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        onlyRemoveTypeImports: true,
      },
    ],
    ['@babel/preset-react', { runtime: 'automatic', useBuiltIns: true }],
    'linaria/babel',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: { version: 3, proposals: true },
        helpers: true,
        regenerator: true,
      },
    ],
    'dynamic-import-node',
    ['const-enum', { transform: 'constObject' }],
  ],
});
