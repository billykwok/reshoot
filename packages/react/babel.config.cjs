module.exports = {
  targets: { esmodules: true },
  presets: [
    ['@babel/preset-env', { targets: { esmodules: true }, bugfixes: true }],
    ['@babel/preset-typescript', { isTSX: true, allExtensions: true }],
    ['@babel/preset-react', { runtime: 'automatic', useBuiltIns: true }],
    '@linaria',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      { corejs: { version: 3 }, helpers: true },
    ],
  ],
  compact: true,
  minified: true,
};
