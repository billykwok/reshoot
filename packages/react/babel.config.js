module.exports = {
  presets: [
    ['@babel/preset-env', { targets: { esmodules: true }, bugfixes: true }],
    [
      '@babel/preset-typescript',
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        onlyRemoveTypeImports: true,
      },
    ],
    '@babel/preset-react',
    'linaria/babel',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: { version: 3, proposals: true },
        helpers: true,
        regenerator: true,
      },
    ],
    ['const-enum', { transform: 'constObject' }],
  ],
};
