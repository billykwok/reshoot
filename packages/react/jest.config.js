module.exports = {
  displayName: '@reshoot/react',
  snapshotResolver: '<rootDir>/test/snapshotResolver.js',
  setupFilesAfterEnv: ['<rootDir>/test/setup.js'],
  coverageProvider: 'v8',
  verbose: true,
  transform: { '^.+\\.[tj]sx?$': './babel-jest-lerna.js' },
};
