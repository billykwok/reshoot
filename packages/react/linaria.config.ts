export default {
  evaluate: true,
  displayName: process.env.NODE_ENV === 'development',
};
