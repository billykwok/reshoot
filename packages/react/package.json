{
  "name": "@reshoot/react",
  "description": "A React component that lazily renders responsive images with LQIPs",
  "version": "1.6.4",
  "type": "commonjs",
  "exports": {
    "import": "./lib/es/index.js",
    "require": "./lib/index.js"
  },
  "main": "./lib/index.js",
  "types": "./types/index.d.ts",
  "files": [
    "lib",
    "types"
  ],
  "sideEffects": [
    "*.css"
  ],
  "publishConfig": {
    "access": "public"
  },
  "author": "Billy Kwok <opensource@billykwok.me>",
  "license": "MIT",
  "keywords": [
    "reshoot",
    "srcset",
    "image",
    "img",
    "react",
    "responsive",
    "babel"
  ],
  "repository": "https://github.com/billykwok/reshoot",
  "bugs": {
    "url": "https://github.com/billykwok/reshoot/issues"
  },
  "homepage": "https://github.com/billykwok/reshoot/tree/master/packages/react",
  "scripts": {
    "lint": "eslint ./src/**",
    "clean": "rm -rf lib types",
    "build": "yarn clean && rollup -c rollup.config.js && rollup -c rollup.config.js --format es --dir ./lib/es && tsc --project tsconfig.json"
  },
  "peerDependencies": {
    "@babel/runtime-corejs3": "^7.11.2",
    "object-assign": "4.1.1",
    "react": "^16.13.1"
  }
}
