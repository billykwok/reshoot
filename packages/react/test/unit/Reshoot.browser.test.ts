import { describe, beforeEach, afterEach, test, expect } from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';
import { createElement, createRef } from 'react';
import {
  MANUAL,
  OFFLINE,
  ERROR,
  HIDDEN,
  LOADING,
  FADING,
  LOADED,
} from '../../src/state';

describe('Reshoot', () => {
  const download = jest.fn(() => null);
  const baseConfig = {
    src: 'image.jpg',
    srcSet: 'image-480.jpg 480w, image-640.jpg 640w',
    alt: 'Test image',
    width: 100,
    height: 80,
    color: '#fff',
    placeholder: 'data:jpeg;base64,1shfk241jkw',
  };
  const useLoadingState = jest.fn();
  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
  const setState = jest.fn((callback: () => number): void => {});

  beforeEach(() => {
    jest.doMock('../../src/utils/isBrowser', () => ({
      __esModule: true,
      default: true,
    }));
    jest.doMock('../../src/hooks/useLoadingState', () => ({
      __esModule: true,
      default: useLoadingState,
    }));
    jest.doMock('../../src/hooks/useIntersection', () => ({
      __esModule: true,
      default: () => null,
    }));
    jest.doMock('../../src/hooks/useDownload', () => ({
      __esModule: true,
      default: () => download,
    }));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('render valid input correctly in state MANUAL', async () => {
    useLoadingState.mockReturnValue([MANUAL, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('render valid input correctly in state OFFLINE', async () => {
    useLoadingState.mockReturnValue([OFFLINE, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('render valid input correctly in state ERROR', async () => {
    useLoadingState.mockReturnValue([ERROR, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();

    const children = dom.getElementsByTagName('button');
    expect(children).toHaveLength(1);

    fireEvent.click(children.item(0));
    expect(download).toHaveBeenCalledTimes(1);
    expect(dom).toMatchSnapshot();
  });

  test('render valid input correctly in state HIDDEN', async () => {
    useLoadingState.mockReturnValue([HIDDEN, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('render valid input correctly in state LOADING', async () => {
    useLoadingState.mockReturnValue([LOADING, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('render valid input correctly in state FADING', async () => {
    useLoadingState.mockReturnValue([FADING, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();

    const childrens = dom.getElementsByTagName('div');
    expect(childrens).toHaveLength(1);

    fireEvent.animationEnd(childrens.item(0));
    expect(setState).toHaveBeenCalledTimes(1);
    expect(setState.mock.calls[0][0]()).toEqual(LOADED);
  });

  test('render valid input correctly in state LOADED', async () => {
    useLoadingState.mockReturnValue([LOADED, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });
  test('customized options for image link', async () => {
    useLoadingState.mockReturnValue([LOADED, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: {
          ...baseConfig,
          color: '#fff',
          placeholder: 'placeholder.jpg',
        },
        ref,
        target: '_blank',
        href: 'https://example.com',
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('do not render preview when placeholder is falsy', async () => {
    useLoadingState.mockReturnValue([LOADING, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: { ...baseConfig, color: '#fff', placeholder: false },
        ref,
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('customized options for normal image', async () => {
    useLoadingState.mockReturnValue([LOADED, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: { ...baseConfig, color: '#fff' },
        ref,
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('aspect ratio can be overridden', async () => {
    useLoadingState.mockReturnValue([LOADED, setState]);
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: { ...baseConfig, aspectRatio: 2 },
        ref,
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });
});
