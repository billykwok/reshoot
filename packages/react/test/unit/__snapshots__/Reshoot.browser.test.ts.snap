// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reshoot aspect ratio can be overridden 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.5;"
>
  <img
    height="80"
    src="image.jpg"
    srcset="image-480.jpg 480w, image-640.jpg 640w"
    width="100"
  />
</div>
`;

exports[`Reshoot customized options for image link 1`] = `
<a
  class="ax4jui0"
  data-testid="123"
  href="https://example.com"
  style="color: rgb(255, 255, 255); --r: 0.8;"
  target="_blank"
>
  <img
    height="80"
    src="image.jpg"
    srcset="image-480.jpg 480w, image-640.jpg 640w"
    width="100"
  />
</a>
`;

exports[`Reshoot customized options for normal image 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <img
    height="80"
    src="image.jpg"
    srcset="image-480.jpg 480w, image-640.jpg 640w"
    width="100"
  />
</div>
`;

exports[`Reshoot do not render preview when placeholder is falsy 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
/>
`;

exports[`Reshoot render valid input correctly in state ERROR 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <div>
    <img
      alt=""
      height="80"
      loading="lazy"
      src="data:jpeg;base64,1shfk241jkw"
      width="100"
    />
  </div>
  <button>
    <svg
      viewBox="0 0 100 100"
    >
      <path
        d="M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20"
      />
    </svg>
    <h6>
      Fail to load
    </h6>
    <p>
      Click to reload
    </p>
  </button>
</div>
`;

exports[`Reshoot render valid input correctly in state ERROR 2`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <div>
    <img
      alt=""
      height="80"
      loading="lazy"
      src="data:jpeg;base64,1shfk241jkw"
      width="100"
    />
  </div>
  <button>
    <svg
      viewBox="0 0 100 100"
    >
      <path
        d="M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20"
      />
    </svg>
    <h6>
      Fail to load
    </h6>
    <p>
      Click to reload
    </p>
  </button>
</div>
`;

exports[`Reshoot render valid input correctly in state FADING 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <img
    height="80"
    src="image.jpg"
    srcset="image-480.jpg 480w, image-640.jpg 640w"
    width="100"
  />
  <div
    class="ajkm71e"
  >
    <img
      alt=""
      height="80"
      loading="lazy"
      src="data:jpeg;base64,1shfk241jkw"
      width="100"
    />
  </div>
</div>
`;

exports[`Reshoot render valid input correctly in state HIDDEN 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <div>
    <img
      alt=""
      height="80"
      loading="lazy"
      src="data:jpeg;base64,1shfk241jkw"
      width="100"
    />
  </div>
</div>
`;

exports[`Reshoot render valid input correctly in state LOADED 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <img
    height="80"
    src="image.jpg"
    srcset="image-480.jpg 480w, image-640.jpg 640w"
    width="100"
  />
</div>
`;

exports[`Reshoot render valid input correctly in state LOADING 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <div>
    <img
      alt=""
      height="80"
      loading="lazy"
      src="data:jpeg;base64,1shfk241jkw"
      width="100"
    />
  </div>
</div>
`;

exports[`Reshoot render valid input correctly in state MANUAL 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <div>
    <img
      alt=""
      height="80"
      loading="lazy"
      src="data:jpeg;base64,1shfk241jkw"
      width="100"
    />
  </div>
  <button>
    <svg
      viewBox="0 0 100 100"
    >
      <path
        d="M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20"
      />
    </svg>
    <h6>
      Not autoloaded in slow network
    </h6>
    <p>
      Click to reload
    </p>
  </button>
</div>
`;

exports[`Reshoot render valid input correctly in state OFFLINE 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <div>
    <img
      alt=""
      height="80"
      loading="lazy"
      src="data:jpeg;base64,1shfk241jkw"
      width="100"
    />
  </div>
  <button>
    <svg
      viewBox="0 0 100 100"
    >
      <path
        d="M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20"
      />
    </svg>
    <h6>
      Browser is offline
    </h6>
    <p>
      Click to reload
    </p>
  </button>
</div>
`;
